[Merged by Bors] - feat: continuity of the continuous functional calculus in each variable

j-loreaux opened the pull request:
---

- [ ] depends on: #24860

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ADedecker commented on Mathlib/Analysis/CStarAlgebra/ContinuousFunctionalCalculus/Continuity.lean:
```diff
@@ -0,0 +1,326 @@
+theorem tendsto_cfc_fun {l : Filter X} (F : X → R → R) (f : R → R) (a : A)
+    (h_tendsto : TendstoUniformlyOn F f l (spectrum R a))
+    (hF : ∀ x, ContinuousOn (F x) (spectrum R a) := by cfc_cont_tac)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac) :
```
`hf` is not needed, right ?
(Got distracted while thinking about the other PR)

ADedecker commented on Mathlib/Analysis/CStarAlgebra/ContinuousFunctionalCalculus/Continuity.lean:
```diff
@@ -0,0 +1,326 @@
+
+/-- `cfcHomSuperset` is continuous in the variable `a : A` when `s : Set 𝕜` is compact and `a`
+varies over elements whose spectrum is contained in `s`, all of which satisfy the predicate `p`. -/
+theorem continuous_cfcHomSuperset_left
+    [TopologicalSpace X] {s : Set 𝕜} (hs : IsCompact s) (f : C(s, 𝕜))
+    (a : X → A) (ha_cont : Continuous a) (ha : ∀ x, spectrum 𝕜 (a x) ⊆ s)
+    (ha' : ∀ x, p (a x) := by cfc_tac) :
+    Continuous (fun x ↦ cfcHomSuperset (ha' x) (ha x) f) := by
+  open scoped ContinuousFunctionalCalculus in
+  have : CompactSpace s := by rwa [isCompact_iff_compactSpace] at hs
+  induction f using ContinuousMap.induction_on_of_compact with
+  | const r =>
+    have : ContinuousMap.const s r = algebraMap 𝕜 C(s, 𝕜) r := rfl
+    simpa only [this, AlgHomClass.commutes] using continuous_const
+  | id =>
+    simp only [cfcHomSuperset_id]
+    fun_prop
+  | star_id =>
+    simp only [map_star, cfcHomSuperset_id]
+    fun_prop
+  | add f g hf hg => simpa using hf.add hg
+  | mul f g hf hg => simpa using hf.mul hg
+  | frequently f hf =>
+    apply continuous_of_uniform_approx_of_continuous
+    rw [Metric.uniformity_basis_dist_le.forall_iff (by aesop)]
+    intro ε hε
+    simp only [Set.mem_setOf_eq, dist_eq_norm]
+    obtain ⟨g, hg, g_cont⟩ := frequently_iff.mp hf (Metric.closedBall_mem_nhds f hε)
+    simp only [Metric.mem_closedBall, dist_comm g, dist_eq_norm] at hg
+    refine ⟨_, g_cont, fun x ↦ ?_⟩
+    rw [← map_sub, cfcHomSuperset_apply]
+    rw [isometry_cfcHom (R := 𝕜) _ (ha' x) |>.norm_map_of_map_zero (map_zero (cfcHom (ha' x)))]
+    rw [ContinuousMap.norm_le _ hε.le] at hg ⊢
+    aesop
```
Here's two ways to do this, which are ultimately the same but phrased differently.

1) Consider the family `fun x ↦ cfcHomSuperset (ha' x) (ha x)`. It is equicontinuous thanks to [NormedSpace.equicontinuous_TFAE](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/NormedSpace/OperatorNorm/NormedSpace.html#NormedSpace.equicontinuous_TFAE). Thus, by [Equicontinuous.isClosed_setOf_tendsto](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Equicontinuity.html#Equicontinuous.isClosed_setOf_tendsto), it suffices to check pointwise convergence on a dense subset of `C(s, 𝕜)`. So you check on polynomials and their conjugates, and you're done.

2) The linear map `C(s, 𝕜) →ₗ[𝕜] (X →ᵤ A)` given by `fun f ↦ fun x ↦ cfcHomSuperset (ha' x) (ha x) f` is continuous because it has norm less than one. When `f` is a star-polynomial this lands in continuous functions, hence this is the case everywhere thanks to [UniformFun.isClosed_setOf_continuous](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/UniformConvergenceTopology.html#UniformFun.isClosed_setOf_continuous)

j-loreaux replied:
I've thought about this a bit, but I don't think it's really worth it. In both cases, there's too much bundling / unbundling to be worth the trouble. I even tried implementing the first approach, but it wasn't going to be any shorter. For instance you have to first bundle `fun x ↦ cfcHomSuperset (ha' x) (ha x)` into a map `X → C(s, 𝕜) →L[𝕜] A`, instead of the current bundling `X → C(s, 𝕜) →⋆ₐ[𝕜] A`. Either that, or we would have to add a dedicated `cfcLSuperset` variant which is a melding of `cfcL` and `cfcHom`. Mind you, this is only a few lines, but there's lots of little annoyances like this, like proving equicontinuity.

That said, I understand that the equicontinuity approach is more mathematically sophisticated and gives a better understanding of what's going on. If you want to implement it, I won't complain, but I don't really think it's worth the effort.

dupuisf commented on Mathlib/Analysis/CStarAlgebra/ContinuousFunctionalCalculus/Continuity.lean:
```diff
@@ -0,0 +1,352 @@
+    [PartialOrder A] [StarOrderedRing A] [NonnegSpectrumClass ℝ A]
+    [T2Space A] [IsTopologicalRing A]
+
+open scoped NNReal in
+/-- A version of `continuous_cfcₙ` over `ℝ≥0` instead of `RCLike 𝕜`. -/
+theorem continuous_cfcₙ_nnreal [TopologicalSpace X] (s : Set ℝ≥0)
```
Is it easy to get `continuousOn` versions of this as well?

j-loreaux commented:
@dupuisf I've implemented a number of changes:

1. For the continuity in the function argument, I've removed autoparams because we now take `∀ᶠ` hypotheses, and so I've also switched explicitness of some arguments. I also added `ContinuousWithinAt` variants and friends. This is done in both the unital and non-unital case.
2. For the continuity in the operator argument, I've added all the variants you might want (i.e., `Tendsto` etc.), derived from `continuousOn_cfc`, but for the moment I only did this in the unital `RCLike 𝕜` case.
3. I proved `continuousOn_cfc_nnreal` as you mentioned, so it is possible to add all these variations for the `ℝ≥0` case too. Of course, all this is doable for the non-unital case also.

If you approve of all this, I will update the `ℝ≥0` and non-unital cases to match.

dupuisf commented on Mathlib/Topology/ContinuousMap/ContinuousMapZero.lean:
```diff
@@ -52,6 +52,15 @@ instance instContinuousMapClass : ContinuousMapClass C(X, R)₀ X R where
+/-- not marked as an instance because it would be a bad one in general, but it can
+be useful when working with `ContinuousMapZero` and the non-unital continuous
+functional calculus. -/
+def _root_.Set.zeroOffFactMem {X : Type*} [Zero X] (s : Set X) [Fact (0 ∈ s)] :
+    Zero s where
```
```suggestion
def _root_.Set.zeroOfFactMem {X : Type*} [Zero X] (s : Set X) [Fact (0 ∈ s)] :
    Zero s where
```

dupuisf reviewed (commented):
Looks good, thanks!

bors d+

j-loreaux commented on Mathlib/Topology/ContinuousMap/ContinuousMapZero.lean:
```diff
@@ -52,6 +52,15 @@ instance instContinuousMapClass : ContinuousMapClass C(X, R)₀ X R where
+    Zero s where
+  zero := ⟨0, Fact.out⟩
+
+scoped[ContinuousMapZero] attribute [instance] Set.zeroOffFactMem
```
```suggestion
scoped[ContinuousMapZero] attribute [instance] Set.zeroOfFactMem
```

j-loreaux commented:
bors merge
